backend/cn/lib/explain.ml

```diff
@@ -26,6 +26,430 @@ type log_entry =
+      in
+      (* query solver *)
+      ask_solver ctxt.global (Base.List.dedup_and_sort ~compare:LC.compare cs')
+
+
+  (* get a list of constraints that are satisfiable iff candidate could have come from this clause body *)
+  and get_body_constraints
+    (exp : IT.t)
+    (var_def_locs : def_line Sym.Map.t)
+    (candidate : IT.t)
+    (ctxt : C.t)
+    (iargs : (Sym.t * BT.t) list)
+    (term_vals : (IT.t * IT.t) list)
+    =
+    let f var_cands =
+      (* find constraints from checking each variable one at a time *)
+      let accumulate_results acc (v, v_cand) =
+        let@ acc_lcs, acc_var_cands = acc in
+        let@ v_lcs, v_var_cands =
+          get_var_constraints v v_cand acc_var_cands var_def_locs ctxt iargs term_vals
+        in
+        Yes (List.append v_lcs acc_lcs, v_var_cands)
+      in
+      List.fold_left accumulate_results (Yes ([], var_cands)) (Sym.Map.bindings var_cands)
+    in
+    (* use candidate to get terms for FVs in exp *)
+    match get_var_cands exp candidate with
+    | Yes var_cands -> f var_cands
+    | No e -> No e
+    | Error e -> Error e
+    | Unknown e ->
+      let s = Solver.make ctxt.global in
+      let loc = Cerb_location.unknown in
+      let th = !Solver.try_hard in
+      let _ = Solver.try_hard := true in
+      let res =
+        match Solver.ask_solver s [ LC.T (IT.eq_ (exp, candidate) loc) ] with
+        | Sat ->
+          (* not using model to get var cands because it may overconstrain *)
+          Yes ([], Sym.Map.empty)
+        | Unsat -> No !^"Solver returned no at variable assignment stage."
+        | Unknown -> Unknown e
+      in
+      let _ = Solver.try_hard := th in
+      res
+
+
+  and get_var_constraints
+    (v : Sym.t)
+    (v_cand : IT.t)
+    (var_cands : IT.t Sym.Map.t)
+    (var_def_locs : def_line Sym.Map.t)
+    (ctxt : C.t)
+    (iargs : (Sym.t * BT.t) list)
+    (term_vals : (IT.t * IT.t) list)
+    =
+    let loc = Cerb_location.unknown in
+    (* find def of x *)
+    match Sym.Map.find_opt v var_def_locs with
+    | None ->
+      (match (Sym.Map.find_opt v ctxt.logical, Sym.Map.find_opt v ctxt.computational) with
+       | Some (Value it, _), _ ->
+         get_body_constraints it var_def_locs v_cand ctxt iargs term_vals
+       | _, Some (Value it, _) ->
+         get_body_constraints it var_def_locs v_cand ctxt iargs term_vals
+       (* TODO: logical vs computational *)
+       (* TODO: BaseType case *)
+       | _ ->
+         let f (s, _) = Sym.equal s v in
+         (match List.find_opt f iargs with
+          | Some _ -> Yes ([], var_cands)
+          | _ -> Unknown (!^"Could not find variable definition line for" ^^^ Sym.pp v)))
+    | Some line ->
+      (match line with
+       (* recurse with x's definition *)
+       | DefineL ((_, t), _) ->
+         get_body_constraints t var_def_locs v_cand ctxt iargs term_vals
+       | ResourceL ((_, (p, _)), _) ->
+         (match p with
+          | P psig ->
+            (match psig.name with
```

Comment:
You can match directly on the record and field to reduce the nesting and make the code more readable:
```ocaml
| P { name = Owned _; (* remaining fields *) } ->
  (* code *)
| P { name = PName name; (* remaining fields *) } ->
  (* other code *)
| Q _ -> 
  (* more code *)
```